proxy/resolver/resolver.go

```diff
@@ -0,0 +1,219 @@
+package resolver
+
+import (
+	"net/url"
+	"time"
+	"log"
+	"sync"
+	"database/sql"
+	"fmt"
+	_ "github.com/mattn/go-sqlite3"
+)
+
+// -- Public API
+
+// Configuration
+type Config struct {
+	DbDriver string        // < Database engine, eg. sqlite3 or Postgres
+	DbConnectString string // < Driver specific connection string
+}
+
+// Running container instance that can be proxied to.
+// (used internally for invalid containers also)
+type Container struct {
+	Subdomain string   // < Subdomain the container is routed under
+	TargetUrl *url.URL // < URL of the running container instance
+
+	// Internal
+	err    error     // < If not `nil`, then the container is not reachable
+	expiry time.Time // < Cached data is valid until this time
+}
+
+// Initialize the resolver, must be called before Resolve()
+func Initialize(config *Config) error {
+	var err error
+	db, err = sql.Open(config.DbDriver, config.DbConnectString)
+	if err != nil {
+		return err
+	}
+
+	for i := 0; i < 10; i++ {
+		go databaseWorker()
+	}
+	go resolveWorker()
+
+	return nil
+}
+
+// Resolves a container for a given public ID. May result in a database lookup
+// and take a while to return. Returns `nil` if no matching container instance
+// is found.
+func Resolve(id string) (*Container, error) {
+	container := getCachedContainer(id)
+
+	if container == nil {
+		response := make(chan *Container)
+		resolveRequests <- resolveRequest {
+			id: id,
+			response: response,
+		}
+		container = <-response
+	}
+
+	if container.err != nil {
+		return nil, container.err
+	}
+
+	return container, nil
+}
+
+// -- Implementation
+
+// An in-flight container resolve request. The resolved container is written
+// through response. Sort of like a "promise" of Container. Always returns a
+// pointer to a container, but it may have `err != nil`
+type resolveRequest struct {
+	id string                  // < Request input
+	response chan<- *Container // < Found container
+}
+
+// -- Container cache
+// Containers are cached for some amount of time for more responsive user experience
+// and less database load.
+
+var containerCache = make(map[string]*Container)
+var cacheMutex sync.RWMutex
+
+// Try to find a container in the cache, returns `nil` if not present or
+// if the cached data is outdated.
+func getCachedContainer(id string) *Container {
+	cacheMutex.RLock()
+	container, ok := containerCache[id]
+	cacheMutex.RUnlock()
+
+	if ok && time.Now().Before(container.expiry) {
+		return container
+	} else {
+		return nil
+	}
+}
+
+// -- Database resolver
+
+var db *sql.DB
+
+// Pending requests to the database
+var databaseRequests = make(chan resolveRequest)
+
+// Query the database and create a container instance
+func fetchContainerFromDatabase(id string) (*Container, error) {
+	row := db.QueryRow("SELECT subdomain, url FROM containers WHERE subdomain = ?", id)
```

Comment:
Voi tulla ongelmia Postgres:in kanssa, nodessa postgres käytti $1,$2... '?','?' sijaan

Reply:
Will test when I can merge Postgres here